R/fwrite.R

```diff
@@ -0,0 +1,61 @@
+fwrite <- function(x, file.path, append = FALSE, quote = TRUE,
+                   sep = ",", eol = "\n", col.names = TRUE, qmethod = "double",
+                   block.size = 10000) {
+  
+  # validate arguments
+  stopifnot(is.data.frame(x))
+  stopifnot(ncol(x) > 0)
+  
+  stopifnot(length(quote) == 1 && class(quote) == "logical")
+  stopifnot(length(sep) == 1 && class(sep) == "character" && nchar(sep) == 1)
+  stopifnot(length(eol) == 1 && class(eol) == "character")
+  stopifnot(length(qmethod) == 1 && qmethod %in% c("double", "escape"))
+  stopifnot(length(col.names) == 1 && class(col.names) == "logical")
+  stopifnot(length(append) == 1 && class(append) == "logical")
+  stopifnot(length(block.size) == 1 && block.size > 0)
+  
+  quoted_cols <- rep(quote, ncol(x))
+  
+  # special case: single-column data.frame, doing x[block_begin:block_end,]
+  # for such data frame gives a vector
+  if (!is.data.table(x) && ncol(x) == 1) x <- as.data.table(x)
+  
+  # write header row separately for correct quoting of row names
+  if (col.names && !append) {
+    .Call(Cwritefile, as.list(names(x)), file.path, sep, eol, quoted_cols, qmethod == "escape", append)
+    append <- TRUE
+  }
+  
+  # handle empty x
+  if (nrow(x) == 0) return()
+  
+  # determine from column types, which ones should be quoted
+  if (quote) {
+    column_types <- sapply(x, class)
+    quoted_cols <- column_types %chin% c('character', 'factor')
+  }
+  
+  # write in blocks of given size to avoid generating full copies
+  # of columns in memory
+  block_begin <- 1
+  
+  repeat {
+    block_end <- min(block_begin+(block.size-1), nrow(x))
+    
+    dt_block <- x[c(block_begin:block_end),]
+    
+    # convert data.frame row block to a list of columns
+    col_list <- lapply(dt_block, function(column) {
+      str_col <- as.character(column)
```

Comment:
I'm not 100% sure I'm interpreting the output of `profvis` correctly, but I'm pretty sure this here is by far the biggest bottleneck on the code as it stands:

![screenshot from 2016-03-28 09 56 00](https://cloud.githubusercontent.com/assets/7606389/14079430/a694a626-f4cb-11e5-8691-34477c245d51.png)

I think all the time spent on `Cwritefile` is in the tiny slices between `lapply` blocks (AFAICT `profvis` has no way of recording anything about C code, not even that the compiler is currently in C).

If that's correct, it seems there's ample room for improvement by porting the formatting step into C.

----
Here's the code I used for that:

```
setwd("~/Desktop/")
library(profvis)
library(data.table)
NN <- 1e6
set.seed(51423)
testDT <- 
  data.table(str1=sample(sprintf("%010d",1:NN)), #ID field 1
             str2=sample(sprintf("%09d",1:NN)),  #ID field 2
             #varying length string field--think names/addresses, etc.
             str3=replicate(NN,paste0(sample(LETTERS,sample(10:30,1),T),
                                      collapse="")),
             #factor-like string field with 50 "levels"
             str4=sprintf("%05d",sample(sample(1e5,50),NN,T)),
             #factor-like string field with 17 levels, varying length
             str5=sample(replicate(17,paste0(sample(LETTERS,
                                                    sample(15:25,1),T),
                                             collapse="")),NN,T),
             #lognormally distributed numeric
             num1=exp(rnorm(NN,mean=6.5,sd=1.5)),
             #3 binary strings
             str6=sample(c("Y","N"),NN,T),
             str7=sample(c("M","F"),NN,T),
             str8=sample(c("B","W"),NN,T),
             #right-skewed integer
             int1=ceiling(rexp(NN)),
             #dates by month
             dat1=sample(seq(from=as.Date("2005-12-31"),
                             to=as.Date("2015-12-31"),by="month"),
                         NN,T),
             dat2=sample(seq(from=as.Date("2005-12-31"),
                             to=as.Date("2015-12-31"),by="month"),
                         NN,T),
             num2=exp(rnorm(NN,mean=6,sd=1.5)),
             #date by day
             dat3=sample(seq(from=as.Date("2015-06-01"),
                             to=as.Date("2015-07-15"),by="day"),
                         NN,T),
             #lognormal numeric that can be positive or negative
             num3=(-1)^sample(2,NN,T)*exp(rnorm(NN,mean=6,sd=1.5)))

profvis(fwrite(testDT, "test.csv"))
```